gymnasium/envs/external/gym_env.py

```diff
@@ -0,0 +1,212 @@
+            )
+
+        if make_kwargs is None:
+            make_kwargs = {}
+
+        if env is not None:
+            self.gym_env = env
+        elif env_id is not None:
+            self.gym_env = gym.make(env_id, **make_kwargs)
+        else:
+            raise gymnasium.error.MissingArgument(
+                "Either env_id or env must be provided to create a legacy gym environment."
+            )
+        self.gym_env = _strip_default_wrappers(gym.make(env_id, **make_kwargs))
+
+        self.observation_space = _convert_space(self.gym_env.observation_space)
+        self.action_space = _convert_space(self.gym_env.action_space)
+
+        self.metadata = getattr(self.gym_env, "metadata", {"render_modes": []})
+        self.render_mode = self.gym_env.render_mode
+        self.reward_range = getattr(self.gym_env, "reward_range", None)
+        self.spec = getattr(self.gym_env, "spec", None)
+
+    def reset(
+        self, seed: Optional[int] = None, options: Optional[dict] = None
+    ) -> Tuple[ObsType, dict]:
+        """Resets the environment.
+
+        Args:
+            seed: the seed to reset the environment with
+            options: the options to reset the environment with
+
+        Returns:
+            (observation, info)
+        """
+        super().reset(seed=None)  # We don't need the seed inside gymnasium
+        # Options are ignored
+        return self.gym_env.reset()
+
+    def step(self, action: ActType) -> Tuple[ObsType, float, bool, bool, dict]:
+        """Steps through the environment.
+
+        Args:
+            action: action to step through the environment with
+
+        Returns:
+            (observation, reward, terminated, truncated, info)
+        """
+        return self.gym_env.step(action)
+
+    def render(self):
+        """Renders the environment.
+
+        Returns:
+            The rendering of the environment, depending on the render mode
+        """
+        return self.gym_env.render()
+
+    def close(self):
+        """Closes the environment."""
+        self.gym_env.close()
+
+    def __str__(self):
+        return f"GymEnvironment({self.gym_env})"
+
+    def __repr__(self):
+        return f"GymEnvironment({self.gym_env})"
+
+
+def _strip_default_wrappers(env: "gym.Env") -> "gym.Env":
+    """Strips builtin wrappers from the environment.
+
+    Args:
+        env: the environment to strip builtin wrappers from
+
+    Returns:
+        The environment without builtin wrappers
+    """
+    import gym.wrappers.compatibility  # Cheat because gym doesn't expose it in __init__
+    import gym.wrappers.env_checker  # Cheat because gym doesn't expose it in __init__
+
+    default_wrappers = (
+        gym.wrappers.RenderCollection,
+        gym.wrappers.HumanRendering,
+    )
+    while isinstance(env, default_wrappers):
+        env = env.env
+    return env
+
+
+def _convert_space(space: "gym.Space") -> gymnasium.Space:
```

Comment:
Why are you not using the single dispatch?

Reply:
Initially I had a version with single dispatch, but it crashes the code when running without gym installed - the functions are defined on the top level, so they are always created, and the registration decorator has to read the type annotations. Those type annotations are invalid if gym isn't installed, so single dispatch crashes.

Reply:
(there's probably a way to do some fancy conditional imports that would avoid it, but I figured this is the simpler way)

Reply:
Agreed, this is the simpler way that there should be very few cases where users need to convert types. My only idea is if the function can defined in the environment constructor, does the decorator function still run?

Reply:
If the function was defined inside `__init__`, then I'm pretty sure it wouldn't run automatically, so it'd work. But this is also very cursed, and loses out on the potential benefit of doing single dispatch anyways (you can't register extra functions from the outside anyways), so probably not the best option.

Reply:
Good point, I guess we need to keep the function version then for now until we move to shimmy then gym can be an optional install